test/Stackctl/Config/RequiredVersionSpec.hs

```diff
@@ -0,0 +1,111 @@
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+module Stackctl.Config.RequiredVersionSpec
+  ( spec
+  ) where
+
+import Stackctl.Prelude
+
+import Data.Version
+import Stackctl.Config.RequiredVersion
+import Test.Hspec
+
+data TestCase = TestCase
+  { tcOp :: Maybe Text
+  , tcRequired :: Text
+  , tcVersions :: [([Int], Bool)]
+  }
+
+-- NB. These aren't exhaustive, but I am le tired
+testCases :: [TestCase]
+testCases =
+  [ TestCase
+    Nothing
+    "1.2.3.4"
+    [ ([1, 2, 3, 4], True)
+    , ([1, 2, 3, 3], False)
+    , ([1, 2, 3, 5], False)
+    , ([1, 2, 4], False)
+    ]
+  , TestCase
+    (Just "=")
+    "1.2.3.4"
+    [ ([1, 2, 3, 4], True)
+    , ([1, 2, 3, 3], False)
+    , ([1, 2, 3, 5], False)
+    , ([1, 2, 4], False)
+    ]
+  , TestCase
+    (Just ">=")
+    "1.2.3.4"
+    [ ([1, 2, 3, 4], True)
+    , ([1, 2, 3, 5], True)
+    , ([1, 2, 4], True)
+    , ([1, 2, 3, 3], False)
+    , ([1, 2, 3], False)
+    ]
+  , TestCase
+    (Just ">=")
+    "1.2"
+    [ ([1, 2], True)
+    , ([1, 2, 1], True)
+    , ([1, 1, 9], False)
+    , ([1, 1, 9, 11], False)
+    , ([1, 0], False)
+    ]
+  , TestCase
+    (Just "<=")
+    "1.2.3.4"
+    [ ([1, 2, 3, 4], True)
+    , ([1, 2, 3, 3], True)
+    , ([1, 2, 3], True)
+    , ([1, 2, 3, 5], False)
+    , ([1, 2, 3, 4, 1], False)
+    , ([1, 2, 4], False)
+    ]
+  , TestCase
+    (Just "<=")
+    "1.2"
+    [ ([1, 2], True)
+    , ([1, 2, 1], False)
+    , ([1, 1, 9], True)
+    , ([1, 1, 9, 11], True)
+    , ([1, 0], True)
+    ]
+  , TestCase
+    (Just "=~")
+    "1.2.3.4"
+    [([1, 2, 3, 3], False), ([1, 2, 3, 4], True), ([1, 2, 3, 5], False)]
+  , TestCase
+    (Just "=~")
+    "1.2.3"
+    [ ([1, 2, 3], True)
+    , ([1, 2, 3, 1], True)
+    , ([1, 2, 3, 9], True)
+    , ([1, 2, 4], False)
+    , ([1, 2], False)
+    ]
+  ]
```

Comment:
the cases that I think are most impt to test are the `=~`; though you could also just test that new operator itself directly. Otherwise you're relying upon the Ord instance in Data.Version, so I dont think you need to worry about doing too much testing there.